state.go

```diff
@@ -0,0 +1,387 @@
+	ContainerIDs    []string `json:"container_ids"`
+}
+
+// vmStateOnDisk is used to (re)store vm struct on disk
+// XXX stateFormatVersion must be update in case of any changes in this struct.
+type vmStateOnDisk struct {
+	RegisterVM api.RegisterVM `json:"registerVM"`
+	Tokens     []string       `json:"tokens"`
+}
+
+func logContID(containerID string) *logrus.Entry {
+	return proxyLog.WithField("container", containerID)
+}
+
+// On success returns nil, otherwise an error string message.
+func (proxy *proxy) restoreTokens(vm *vm, tokens []string) error {
+	if vm == nil {
+		return fmt.Errorf("vm parameter is nil")
+	}
+
+	for _, token := range tokens {
+		if token == "" {
+			continue
+		}
+
+		token, err := vm.AllocateTokenAs(Token(token))
+		if err != nil {
+			return err
+		}
+
+		proxy.Lock()
+		proxy.tokenToVM[token] = &tokenInfo{
+			state: tokenStateAllocated,
+			vm:    vm,
+		}
+		proxy.Unlock()
+
+		session := vm.findSessionByToken(token)
+		if session == nil {
+			vm.Lock()
+			for token := range vm.tokenToSession {
+				_ = vm.freeTokenUnlocked(token)
+				delete(vm.tokenToSession, token)
+			}
+			vm.Unlock()
+			return fmt.Errorf("unknown token %s", token)
+		}
+
+		// Signal that the process is already started
+		close(session.processStarted)
+	}
+	return nil
+}
+
+// Returns false if it's a clean start (i.e. no state is stored) or restoring
+// failed.
+func (proxy *proxy) restoreAllState() bool {
+	if _, err := os.Stat(storeStateDir); os.IsNotExist(err) {
+		err := os.MkdirAll(storeStateDir, proxyStateDirPerm)
+		if err != nil {
+			proxyLog.Errorf("couldn't create directory %s: %v",
+				storeStateDir, err)
+		}
+		return false
+	}
+
+	fdata, err := ioutil.ReadFile(proxyStateFilePath)
+	if err != nil {
+		proxyLog.Errorf("couldn't read state file %s: %v",
+			proxyStateFilePath, err)
+		return false
+	}
+
+	var proxyState proxyStateOnDisk
+
+	err = json.Unmarshal(fdata, &proxyState)
+	if err != nil {
+		proxyLog.Errorf("couldn't unmarshal %s: %v",
+			proxyStateFilePath, err)
+		return false
+	}
+
+	proxyLog.Debugf("proxyState: %+v", proxyState)
+
+	if len(proxyState.ContainerIDs) == 0 {
+		proxyLog.Warnf("containerIDs list is empty")
+		return false
+	}
+
+	proxyLog.Info("Recovering proxy state from: ", proxyStateFilePath)
+	if proxyState.Version > stateFormatVersion {
+		proxyLog.Errorf("stored state format version (%d) is higher "+
+			"than supported (%d). Aborting", proxyState.Version,
+			stateFormatVersion)
+		return false
+	}
+
+	proxy.socketPath = proxyState.SocketPath
+	proxy.enableVMConsole = proxyState.EnableVMConsole
+
+	for _, containerID := range proxyState.ContainerIDs {
+		go func(contID string) {
+			// ignore failures here but log them inside
+			_ = restoreVMState(proxy, contID)
+		}(containerID)
+	}
+
+	return true
+}
+
+// On success returns nil, otherwise an error string message.
+func (proxy *proxy) storeState() error {
+	proxyStateFilePath := storeStateDir + proxyStateFileName
```

Comment:
If change this function to receive proxy as parameter then I need getter methods for proxy.socketPath, proxy.enableVMConsole and vm[] since they are private.

Reply:
OK, I'll rework this. Thank you.